File: app/src/main/java/ch/threema/logging/ThreemaLogger.java
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2018-2024 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.logging;

import android.util.Log;

import org.slf4j.helpers.MarkerIgnoringBase;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import ch.threema.logging.backend.LogBackend;

/**
 * The Threema logger. It logs all messages using the registered backend(s).
 */
@SuppressWarnings("unused")
public class ThreemaLogger extends MarkerIgnoringBase {
	@NonNull
	private final String tag;
	@NonNull
	private final List<LogBackend> backends = new ArrayList<>();
	@Nullable
	private String prefix = null;


	// Constructors

	ThreemaLogger(@NonNull String logTag, @NonNull LogBackend backend) {
		this.tag = logTag;
		this.backends.add(backend);
	}
	ThreemaLogger(@NonNull String logTag, @NonNull LogBackend[] backends) {
		this.tag = logTag;
		this.backends.addAll(Arrays.asList(backends));
	}
	ThreemaLogger(@NonNull String logTag, @NonNull List<LogBackend> backends) {
		this.tag = logTag;
		this.backends.addAll(backends);
	}


	// Delegate logging to backends

	public void print(@LogLevel int level, @Nullable Throwable throwable, @Nullable String message) {
		if (this.prefix != null) {
			message = this.prefix + ": " + message;
		}
		for (LogBackend backend : this.backends) {
			backend.print(level, this.tag, throwable, message);
		}
	}
	public void print(@LogLevel int level, @Nullable Throwable throwable, @NonNull String messageFormat, Object... args) {
		Throwable extractedThrowable = throwable != null ? throwable : getThrowableCandidate(args);

		if (this.prefix != null) {
			messageFormat = this.prefix + ": " + messageFormat;
		}
		for (LogBackend backend : this.backends) {
			backend.print(level, this.tag, extractedThrowable, messageFormat, args);
		}
	}


	// Set prefix if desired

	public void setPrefix(@Nullable String prefix) {
		this.prefix = prefix;
	}


	// Log levels

	@Override
	public boolean isTraceEnabled() {
		return true;
	}
	@Override
	public boolean isDebugEnabled() {
		return true;
	}
	@Override
	public boolean isInfoEnabled() {
		return true;
	}
	@Override
	public boolean isWarnEnabled() {
		return true;
	}
	@Override
	public boolean isErrorEnabled() {
		return true;
	}


	// Logging calls

	@Override
	public void trace(String msg) {
		this.print(Log.VERBOSE, null, msg);
	}
	@Override
	public void trace(String format, Object arg) {
		this.print(Log.VERBOSE, null, format, arg);
	}
	@Override
	public void trace(String format, Object arg1, Object arg2) {
		this.print(Log.VERBOSE, null, format, arg1, arg2);
	}
	@Override
	public void trace(String format, Object... arguments) {
		this.print(Log.VERBOSE, null, format, arguments);
	}
	@Override
	public void trace(String msg, Throwable t) {
		this.print(Log.VERBOSE, t, msg);
	}

	@Override
	public void debug(String msg) {
		this.print(Log.DEBUG, null, msg);
	}
	@Override
	public void debug(String format, Object arg) {
		this.print(Log.DEBUG, null, format, arg);
	}
	@Override
	public void debug(String format, Object arg1, Object arg2) {
		this.print(Log.DEBUG, null, format, arg1, arg2);
	}
	@Override
	public void debug(String format, Object... arguments) {
		this.print(Log.DEBUG, null, format, arguments);
	}
	@Override
	public void debug(String msg, Throwable t) {
		this.print(Log.DEBUG, t, msg);
	}

	@Override
	public void info(String msg) {
		this.print(Log.INFO, null, msg);
	}
	@Override
	public void info(String format, Object arg) {
		this.print(Log.INFO, null, format, arg);
	}
	@Override
	public void info(String format, Object arg1, Object arg2) {
		this.print(Log.INFO, null, format, arg1, arg2);
	}
	@Override
	public void info(String format, Object... arguments) {
		this.print(Log.INFO, null, format, arguments);
	}
	@Override
	public void info(String msg, Throwable t) {
		this.print(Log.INFO, t, msg);
	}

	@Override
	public void warn(String msg) {
		this.print(Log.WARN, null, msg);
	}
	@Override
	public void warn(String format, Object arg) {
		this.print(Log.WARN, null, format, arg);
	}
	@Override
	public void warn(String format, Object arg1, Object arg2) {
		this.print(Log.WARN, null, format, arg1, arg2);
	}
	@Override
	public void warn(String format, Object... arguments) {
		this.print(Log.WARN, null, format, arguments);
	}
	@Override
	public void warn(String msg, Throwable t) {
		this.print(Log.WARN, t, msg);
	}

	@Override
	public void error(String msg) {
		this.print(Log.ERROR, null, msg);
	}
	@Override
	public void error(String format, Object arg) {
		this.print(Log.ERROR, null, format, arg);
	}
	@Override
	public void error(String format, Object arg1, Object arg2) {
		this.print(Log.ERROR, null, format, arg1, arg2);
	}
	@Override
	public void error(String format, Object... arguments) {
		this.print(Log.ERROR, null, format, arguments);
	}
	@Override
	public void error(String msg, @Nullable Throwable t) {
		this.print(Log.ERROR, t, msg);
	}

	/**
	 * Get the throwable if the last element is a throwable.
	 *
	 * @param argArray the arguments
	 * @return the throwable if the last element is a throwable
	 */
	@Nullable
	private Throwable getThrowableCandidate(@Nullable final Object[] argArray) {
		if (argArray == null || argArray.length == 0) {
			return null;
		}

		final Object lastEntry = argArray[argArray.length - 1];
		if (lastEntry instanceof Throwable) {
			return (Throwable) lastEntry;
		}

		return null;
	}

}

File: app/src/main/java/ch/threema/logging/ThreemaLogger.kt
package ch.threema.logging

import android.util.Log
import ch.threema.logging.backend.LogBackend
import org.slf4j.helpers.MarkerIgnoringBase

class ThreemaLogger(
    private val tag: String,
    private val backends: List<LogBackend>,
) : MarkerIgnoringBase() {

    var prefix: String? = null

    override fun isTraceEnabled() = true

    override fun isDebugEnabled() = true

    override fun isInfoEnabled() = true

    override fun isWarnEnabled() = true

    override fun isErrorEnabled() = true

    override fun trace(msg: String?) {
        print(Log.VERBOSE, null, msg)
    }

    override fun trace(format: String, arg: Any?) {
        print(Log.VERBOSE, format, arrayOf(arg))
    }

    override fun trace(format: String, arg1: Any?, arg2: Any?) {
        print(Log.VERBOSE, format, arrayOf(arg1, arg2))
    }

    override fun trace(format: String, vararg arguments: Any?) {
        print(Log.VERBOSE, format, arrayOf(*arguments))
    }

    override fun trace(msg: String?, t: Throwable?) {
        print(Log.VERBOSE, t, msg)
    }

    override fun debug(msg: String?) {
        print(Log.DEBUG, null, msg)
    }

    override fun debug(format: String, arg: Any?) {
        print(Log.DEBUG, format, arrayOf(arg))
    }

    override fun debug(format: String, arg1: Any?, arg2: Any?) {
        print(Log.DEBUG, format, arrayOf(arg1, arg2))
    }

    override fun debug(format: String, vararg arguments: Any?) {
        print(Log.DEBUG, format, arrayOf(*arguments))
    }

    override fun debug(msg: String?, t: Throwable?) {
        print(Log.DEBUG, t, msg)
    }

    override fun info(msg: String?) {
        print(Log.INFO, null, msg)
    }

    override fun info(format: String, arg: Any?) {
        print(Log.INFO, format, arrayOf(arg))
    }

    override fun info(format: String, arg1: Any?, arg2: Any?) {
        print(Log.INFO, format, arrayOf(arg1, arg2))
    }

    override fun info(format: String, vararg arguments: Any?) {
        print(Log.INFO, format, arrayOf(*arguments))
    }

    override fun info(msg: String?, t: Throwable?) {
        print(Log.INFO, t, msg)
    }

    override fun warn(msg: String?) {
        print(Log.WARN, null, msg)
    }

    override fun warn(format: String, arg: Any?) {
        print(Log.WARN, format, arrayOf(arg))
    }

    override fun warn(format: String, arg1: Any?, arg2: Any?) {
        print(Log.WARN, format, arrayOf(arg1, arg2))
    }

    override fun warn(format: String, vararg arguments: Any?) {
        print(Log.WARN, format, arrayOf(*arguments))
    }

    override fun warn(msg: String?, t: Throwable?) {
        print(Log.WARN, t, msg)
    }

    override fun error(msg: String?) {
        print(Log.ERROR, null, msg)
    }

    override fun error(format: String, arg: Any?) {
        print(Log.ERROR, format, arrayOf(arg))
    }

    override fun error(format: String, arg1: Any?, arg2: Any?) {
        print(Log.ERROR, format, arrayOf(arg1, arg2))
    }

    override fun error(format: String, vararg arguments: Any?) {
        print(Log.ERROR, format, arrayOf(*arguments))
    }

    override fun error(msg: String?, t: Throwable?) {
        print(Log.ERROR, t, msg)
    }

    private fun print(@LogLevel level: Int, throwable: Throwable?, message: String?) {
        var message = message
        if (prefix != null) {
            message = "$prefix: $message"
        }
        backends.forEach { backend ->
            backend.print(level, tag, throwable, message)
        }
    }

    private fun print(@LogLevel level: Int, messageFormat: String, args: Array<Any?>) {
        var messageFormat = messageFormat
        if (prefix != null) {
            messageFormat = "$prefix: $messageFormat"
        }
        val extractedThrowable = args.lastOrNull() as? Throwable
        backends.forEach { backend ->
            backend.print(level, tag, extractedThrowable, messageFormat, args)
        }
    }
}
